File: internal/server/cli_native.go
package server

import (
	"encoding/json"
	"errors"
	"net/http"
	"sort"
	"strings"

	"github.com/cfpperche/picode/internal/climemory"
	"github.com/cfpperche/picode/internal/climodels"
	"github.com/cfpperche/picode/internal/clisettings"
	"github.com/cfpperche/picode/internal/store"
)

// Native settings and native memory for the guest agent CLIs (ADR-0163).
// Pi is not routed here: it keeps `/api/pi-settings`, its own layers and its
// own trust rules (ADR-0101). Memory does answer for Pi, because the honest
// answer there is "Pi has none" and the pane needs to say it.

func registerCLINativeRoutes(mux Registrar, deps Deps) {
	mux.HandleFunc("GET /api/cli-settings", handleCLISettingsGet(deps))
	mux.HandleFunc("PATCH /api/cli-settings", handleCLISettingsPatch(deps))
	mux.HandleFunc("GET /api/cli-models", handleCLIModelsGet(deps))
	mux.HandleFunc("GET /api/cli-memory", handleCLIMemoryGet(deps))
	mux.HandleFunc("GET /api/cli-memory/item", handleCLIMemoryRead(deps))
	mux.HandleFunc("PUT /api/cli-memory/item", handleCLIMemoryWrite(deps))
	mux.HandleFunc("DELETE /api/cli-memory/item", handleCLIMemoryDelete(deps))
}

// nativePaths resolves the workspace folder these drivers read and write. A
// stale workspace id drops the workspace layer rather than failing the pane,
// matching how connectors behave (ADR-0150).
func nativePaths(deps Deps, workspaceID string) (string, error) {
	if workspaceID == "" || deps.Store == nil {
		return "", nil
	}
	ws, err := deps.Store.GetWorkspace(workspaceID)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			return "", nil
		}
		return "", err
	}
	return ws.Path, nil
}

func handleCLISettingsGet(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		cli := r.URL.Query().Get("cli")
		if clisettings.For(cli) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no settings editor for "+cliLabel(cli))
			return
		}
		cwd, err := nativePaths(deps, r.URL.Query().Get("workspace"))
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		rep, err := clisettings.Read(cli, clisettings.Paths{Cwd: cwd})
		if err != nil {
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		writeJSON(w, http.StatusOK, rep)
	}
}

// handleCLIModelsGet asks a CLI which models it can reach, for the role
// matrix's picker (ADR-0181). It runs the vendor's own read-only command in the
// workspace being edited, because the answer depends on the directory: a
// project that disables a provider gets a shorter list than the folder next to
// it (measured 2026-09-22, omp 18.2.8). The pane calls this when a picker
// opens, never on mount — the probe costs a subprocess.
func handleCLIModelsGet(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		cli := r.URL.Query().Get("cli")
		if !climodels.Supports(cli) {
			writeErr(w, http.StatusBadRequest, "PiCode cannot ask "+cliLabel(cli)+" which models it has")
			return
		}
		cwd, err := nativePaths(deps, r.URL.Query().Get("workspace"))
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		rep, err := climodels.Read(r.Context(), cli, cwd)
		if err != nil {
			// The vendor's own words, not a status code: a CLI that is not
			// installed, not signed in, or slow says so differently each time.
			writeErr(w, http.StatusBadGateway, err.Error())
			return
		}
		writeJSON(w, http.StatusOK, rep)
	}
}

type cliSettingsPatchReq struct {
	CLI         string         `json:"cli"`
	WorkspaceID string         `json:"workspaceId"`
	Scope       string         `json:"scope"`
	Set         map[string]any `json:"set"`
	Reset       []string       `json:"reset"`
	Revision    string         `json:"revision"`
	Force       bool           `json:"force"`
}

func handleCLISettingsPatch(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req cliSettingsPatchReq
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeErr(w, http.StatusBadRequest, "invalid request body")
			return
		}
		if clisettings.For(req.CLI) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no settings editor for "+cliLabel(req.CLI))
			return
		}
		cwd, err := nativePaths(deps, req.WorkspaceID)
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		p := clisettings.Paths{Cwd: cwd}
		err = clisettings.Apply(req.CLI, p, clisettings.Patch{
			Scope:    req.Scope,
			Set:      req.Set,
			Reset:    req.Reset,
			Revision: req.Revision,
			Force:    req.Force,
		})
		switch {
		case errors.Is(err, clisettings.ErrStale):
			// The CLI writes this file too. 409 is the pane's cue to re-read
			// and show what changed before offering Replace.
			writeErr(w, http.StatusConflict, err.Error())
			return
		case err != nil && clisettings.IsParseError(err):
			writeErr(w, http.StatusConflict, err.Error())
			return
		case err != nil:
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		// The file stays authoritative; this only invalidates open views.
		if deps.Feed != nil {
			deps.Feed.Ephemeral("cli.settings", map[string]any{"cli": req.CLI, "scope": req.Scope})
		}
		rep, err := clisettings.Read(req.CLI, p)
		if err != nil {
			writeJSON(w, http.StatusOK, map[string]any{"ok": true})
			return
		}
		writeJSON(w, http.StatusOK, rep)
	}
}

// memoryPaths wires the memory driver to the settings driver, so a store the
// user moved in their CLI's own settings is found where they moved it.
func memoryPaths(cwd string) climemory.Paths {
	return climemory.Paths{
		Cwd: cwd,
		Setting: func(cli, key string) (string, bool) {
			// The same workspace the memory request names, so a key set in the
			// workspace layer is the one that wins here too.
			rep, err := clisettings.Read(cli, clisettings.Paths{Cwd: cwd})
			if err != nil {
				return "", false
			}
			for i := len(rep.Layers) - 1; i >= 0; i-- {
				if v, ok := rep.Layers[i].Values[key]; ok {
					if s, isString := v.(string); isString {
						return s, true
					}
				}
			}
			return "", false
		},
	}
}

func handleCLIMemoryGet(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		cli := r.URL.Query().Get("cli")
		if climemory.For(cli) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no memory view for "+cliLabel(cli))
			return
		}
		cwd, err := nativePaths(deps, r.URL.Query().Get("workspace"))
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		p := memoryPaths(cwd)
		rep, err := climemory.Describe(cli, p)
		if err != nil {
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		body := map[string]any{"report": rep}
		// One store's items come back with the report so the pane paints in a
		// single round trip; `scope` picks which.
		scope := r.URL.Query().Get("scope")
		if scope == "" {
			for _, st := range rep.Stores {
				if st.Resolved && st.Exists {
					scope = st.Scope
					break
				}
			}
		}
		if scope != "" {
			body["scope"] = scope
			items, audit, err := climemory.Survey(cli, p, scope)
			if err == nil {
				body["items"] = items
				// The index the CLI loads every session, measured against the
				// limits that CLI applies — a reader cannot see them anywhere
				// else (2026-09-20).
				if audit.Index.Exists {
					body["index"] = audit.Index
				}
				// A store the CLI keeps under version control answers when its
				// content actually changed; Codex is the one that does.
				if audit.History != nil {
					body["history"] = audit.History
				}
			} else {
				body["itemsError"] = err.Error()
			}
		}
		// The pane warns before writing into a store whose CLI is mid-turn.
		body["running"] = runningTerminalsFor(deps, cli)
		writeJSON(w, http.StatusOK, body)
	}
}

func handleCLIMemoryRead(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		cli := q.Get("cli")
		if climemory.For(cli) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no memory view for "+cliLabel(cli))
			return
		}
		cwd, err := nativePaths(deps, q.Get("workspace"))
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		body, err := climemory.Read(cli, memoryPaths(cwd), q.Get("scope"), q.Get("id"))
		if err != nil {
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"id": q.Get("id"), "body": body})
	}
}

type cliMemoryWriteReq struct {
	CLI         string `json:"cli"`
	WorkspaceID string `json:"workspaceId"`
	Scope       string `json:"scope"`
	ID          string `json:"id"`
	Body        string `json:"body"`
}

func handleCLIMemoryWrite(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req cliMemoryWriteReq
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeErr(w, http.StatusBadRequest, "invalid request body")
			return
		}
		if climemory.For(req.CLI) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no memory view for "+cliLabel(req.CLI))
			return
		}
		cwd, err := nativePaths(deps, req.WorkspaceID)
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		if err := climemory.Write(req.CLI, memoryPaths(cwd), req.Scope, req.ID, req.Body); err != nil {
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		announceMemory(deps, req.CLI)
		writeJSON(w, http.StatusOK, map[string]any{"ok": true})
	}
}

func handleCLIMemoryDelete(deps Deps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		cli := q.Get("cli")
		if climemory.For(cli) == nil {
			writeErr(w, http.StatusBadRequest, "PiCode has no memory view for "+cliLabel(cli))
			return
		}
		cwd, err := nativePaths(deps, q.Get("workspace"))
		if err != nil {
			writeErr(w, statusForStore(err), err.Error())
			return
		}
		if err := climemory.Delete(cli, memoryPaths(cwd), q.Get("scope"), q.Get("id")); err != nil {
			writeErr(w, http.StatusBadRequest, err.Error())
			return
		}
		announceMemory(deps, cli)
		writeJSON(w, http.StatusOK, map[string]any{"ok": true})
	}
}

// announceMemory invalidates open views. The event carries the CLI and nothing
// else: memory text never travels on the feed (ADR-0163).
func announceMemory(deps Deps, cli string) {
	if deps.Feed != nil {
		deps.Feed.Ephemeral("cli.memory", map[string]any{"cli": cli})
	}
}

// runningTerminalsFor names the terminals this CLI holds right now, from the
// authoritative presence registry (ADR-0062), so the pane can say who else is
// writing before the reader edits a memory file. Presence, never a guess from
// the terminal's name.
func runningTerminalsFor(deps Deps, cli string) []map[string]string {
	if deps.TermRuntimes == nil {
		return []map[string]string{}
	}
	names := map[string]string{}
	if deps.Store != nil {
		if rows, err := deps.Store.ListTerminals(); err == nil {
			for _, t := range rows {
				names[t.ID] = t.Name
			}
		}
	}
	out := []map[string]string{}
	for termID, rt := range deps.TermRuntimes.Snapshot() {
		if !strings.EqualFold(rt.CLI, cli) {
			continue
		}
		name := names[termID]
		if name == "" {
			name = "Terminal"
		}
		out = append(out, map[string]string{"id": termID, "name": name})
	}
	sort.Slice(out, func(i, j int) bool { return out[i]["name"] < out[j]["name"] })
	return out
}

func cliLabel(cli string) string {
	if cli == "" {
		return "this CLI"
	}
	return cli
}

File: web/shared/domain/cliNative.js
// Native settings and native memory for the guest agent CLIs (ADR-0163).
//
// Pi is not here. It keeps its own editor, its own API and its own trust rules
// (ADR-0101); this module covers the eight CLIs whose settings PiCode edits by
// writing their own config files, and the memory view every CLI answers for —
// including the ones whose honest answer is "none".

// The CLIs with a settings declaration in internal/clisettings. The server is
// the authority; this list keeps the pane from asking for an editor that does
// not exist, and the test below keeps the two in step.
import { MEMORY_WORDS, readScope, writeScope } from "./scopes.js";
export const NATIVE_SETTINGS_CLIS = ["claude-code", "codex", "grok", "hermes", "opencode", "muse", "agy", "omp"];

export const supportsNativeSettings = (id) => NATIVE_SETTINGS_CLIS.includes(id);

// The route says `layer`, the driver says `scope`. One vocabulary each, mapped
// here rather than in a component, so a link and a request cannot drift.
export function layerToScope(layer) {
  return layer === "project" ? "project" : "user";
}

export function scopeToLayer(scope) {
  return scope === "project" ? "project" : "global";
}

// A workspace layer is named, not generic: a pane bound to a workspace says
// its name where the vendor-declared label says "This workspace" — the rule
// Pi's pane has always followed (label: workspace.name) and Mcps' rows too.
// A pane bound to nothing keeps the generic word, because there is no name to
// say. The replacement is by words, not by scope, so a driver's suffix
// survives: Claude Code's "This workspace (local)" names the checkout its
// uncommitted file is kept in.
export function namedScope(label, workspaceName = "") {
  const name = String(workspaceName || "").trim();
  if (!name) return label;
  // A replacer function, not a replacement string: String.replace interprets
  // `$` sequences ($&, $', $`) in a string replacement, and a workspace name
  // may legally hold them (trimmed, ≤120 chars, no charset rule) — the name
  // must land literally, first occurrence only.
  return String(label || "").replace("This workspace", () => name);
}

// Every layer label in one report, renamed in one pass: a pane feeds its
// layers (settings, models) or memory stores through this once, so the layer
// switcher, the checked radio and the per-row provenance ("From …") all read
// the name. Layers the pane cannot name come back untouched.
export function namedLayers(layers = [], workspaceName = "") {
  const name = String(workspaceName || "").trim();
  if (!name || !Array.isArray(layers)) return layers;
  return layers.map((l) => (
    l && typeof l.label === "string" && l.label.includes("This workspace")
      ? { ...l, label: namedScope(l.label, name) }
      : l
  ));
}

// The layer the pane opens on: the one the route names when the CLI has it,
// otherwise the first layer the CLI declares.
export function defaultLayer(layers = [], routeLayer = "") {
  const wanted = layerToScope(routeLayer);
  if (routeLayer && layers.some((l) => l.scope === wanted)) return routeLayer;
  const first = layers[0];
  return first ? scopeToLayer(first.scope) : "global";
}

// Fields in declaration order, grouped by their declared group. A CLI that
// declares no group gets one unnamed section rather than a header saying
// nothing.
export function groupFields(fields = []) {
  const out = [];
  for (const field of fields) {
    // A row another pane owns (omp's allowed models and hidden providers live
    // in Models) is in the report for the revision, not for this pane.
    if (field.pane) continue;
    const name = field.group || "";
    let group = out.find((g) => g.name === name);
    if (!group) {
      group = { name, fields: [] };
      out.push(group);
    }
    group.fields.push(field);
  }
  return out;
}

// What one row shows: the value in force for this layer, whether this layer is
// the one that sets it, and where it comes from otherwise. `fallback` is the
// CLI's own behaviour when nobody sets the key — the honest answer to "what
// happens if I leave this alone".
export function rowState(field, layers = [], scope = "user") {
  const here = layers.find((l) => l.scope === scope);
  const setHere = !!here && Object.prototype.hasOwnProperty.call(here.values || {}, field.key);
  if (setHere) return { value: here.values[field.key], setHere: true, from: "" };
  // Layers arrive in the vendor's precedence order, last wins; a row this
  // layer does not set inherits from the nearest layer before it that does.
  const index = layers.findIndex((l) => l.scope === scope);
  for (let i = (index < 0 ? layers.length : index) - 1; i >= 0; i -= 1) {
    const layer = layers[i];
    if (layer && Object.prototype.hasOwnProperty.call(layer.values || {}, field.key)) {
      return { value: layer.values[field.key], setHere: false, from: layer.label };
    }
  }
  return { value: undefined, setHere: false, from: "" };
}

// The one-line warning a dangerous choice carries. Danger is named, never
// hidden (docs/benchmarks.md).
export function dangerNote(field, value) {
  if (!field.danger) return "";
  const dangerous = field.danger === "true" ? value === true : String(value) === field.danger;
  if (!dangerous) return "";
  // Each dangerous choice says what it specifically costs. One shared sentence
  // printed the same words twice inside one group (visual review 2026-09-20).
  return field.dangerNote || "This turns off a safety check.";
}

// Values arrive from a config file, so a number field can hold a string and a
// switch can hold anything. Coerce to the kind the field declares, and refuse
// to send a value the field cannot hold.
export function coerce(field, raw) {
  if (field.kind === "bool") return !!raw;
  if (field.kind === "number") {
    const n = Number(String(raw).trim());
    return Number.isFinite(n) ? n : null;
  }
  return String(raw);
}

// Memory item titles come from the file name; the pane shows the CLI's own
// frontmatter kind when a file declares one.
export const MEMORY_KIND_LABELS = {
  user: "About you",
  feedback: "Feedback",
  project: "Project",
  reference: "Reference",
};

export function memoryKindLabel(kind) {
  return MEMORY_KIND_LABELS[kind] || kind || "";
}

// One line for each tier that has nothing to list, plus the action that
// changes it. Chrome carries state and the next action, never an essay
// (.pi/skills/uiux-review).
export function memoryEmptyLine(report, store) {
  if (!report) return { line: "", action: "" };
  if (report.tier === "none" || report.tier === "unknown") {
    return { line: report.note || "", action: "" };
  }
  if (store && !store.resolved) return { line: store.note || "", action: report.toggle ? "settings" : "" };
  if (store && !store.exists) return { line: "Nothing remembered yet.", action: report.toggle ? "settings" : "" };
  return { line: "Nothing remembered yet.", action: report.toggle ? "settings" : "" };
}

export function cliMemoryHash(cli, { workspaceId = "", scope = "" } = {}) {
  const query = new URLSearchParams();
  if (workspaceId) query.set("workspaceId", workspaceId);
  writeScope(query, scope, { keepGlobal: true });
  return "#/clis/" + encodeURIComponent(cli || "") + "/memory" + (query.size ? "?" + query : "");
}

export function cliMemoryLocation(hash = "") {
  const [path, query = ""] = String(hash).replace(/^#/, "").split("?");
  const match = /^\/clis\/([^/]+)\/memory$/.exec(path);
  if (!match) return null;
  let cli = "";
  try {
    cli = decodeURIComponent(match[1]);
  } catch {
    cli = "";
  }
  const params = new URLSearchParams(query);
  const read = readScope(params, MEMORY_WORDS);
  return { view: "clis", pane: "memory", id: cli, workspaceId: params.get("workspaceId") || "", scope: read.value, ...(read.kind ? { scopeKind: read.kind } : {}) };
}

// ── Model roles (ADR-0181) ───────────────────────────────────────────────────
// A CLI that keeps a role → model-selector map reports its rows with the rest
// of its settings, so provenance, "Set here" and "Use inherited" are the same
// code as every other row. What is new is the shape of two values: a selector,
// which carries an optional thinking suffix, and a list.

export const isRoleField = (field) => field?.kind === "role";
export const isListField = (field) => field?.kind === "list";

// A selector is `provider/model-id` with an optional `:level` suffix, and the
// model id may itself carry slashes, dots and an `@upstream` routing suffix
// (omp's `parseModelString`). The split here is **for display only**: the pane
// shows the base in the picker and the level in its own control, and writing
// them back joins the exact two pieces it split, so a value PiCode did not
// understand survives a round trip unchanged.
export function splitSelector(value, levels = []) {
  const text = value === undefined || value === null ? "" : String(value);
  const colon = text.lastIndexOf(":");
  if (colon <= 0) return { base: text, level: "" };
  const suffix = text.slice(colon + 1);
  if (!levels.includes(suffix)) return { base: text, level: "" };
  return { base: text.slice(0, colon), level: suffix };
}

export function joinSelector(base, level) {
  const trimmed = (base || "").trim();
  if (!trimmed) return "";
  return level ? trimmed + ":" + level : trimmed;
}

// A list value arrives from a config file, so it can be a single string (which
// every CLI in this family accepts where a list belongs) or absent.
export function listValue(value) {
  if (Array.isArray(value)) return value.map((v) => String(v));
  if (typeof value === "string" && value !== "") return [value];
  return [];
}

// The role's place in the quick-switch cycle, 1-based, or 0 when it is not in
// it. omp's own hub draws this as `⟳ N` (`model-hub.ts`, "second stop of the
// ctrl+p cycle"), and the pane says the same thing in the same words.
export function cyclePosition(cycle, roleId) {
  return listValue(cycle).indexOf(roleId) + 1;
}

// Every selector any layer already uses, so the picker can offer what this
// machine is actually configured with before the CLI has been asked anything.
export function selectorsInUse(fields = [], layers = []) {
  const seen = new Set();
  for (const field of fields) {
    if (!isRoleField(field)) continue;
    for (const layer of layers) {
      const value = layer?.values?.[field.key];
      if (typeof value === "string" && value && !value.startsWith("@")) seen.add(value);
    }
  }
  return [...seen].sort();
}

// The aliases a role row accepts beside a model: every role is addressable as
// `@name`, and `*` is the CLI's own shorthand for the default role.
export function roleAliases(fields = []) {
  const out = ["*"];
  for (const field of fields) {
    if (isRoleField(field)) out.push("@" + field.key.split(".").pop());
  }
  return out;
}
